flytekit/types/file/file.py

```diff
@@ -300,8 +300,45 @@ def __init__(
         self._remote_source: typing.Optional[str] = None
 
     def __fspath__(self):
-        # This is where a delayed downloading of the file will happen
+        """
+        Define the file path protocol for opening FlyteFile with the context manager,
+        following show two common use cases:
+
+        1. Directly open a FlyteFile with a local path:
+
+        ff = FlyteFile(path=local_path)
+        with open(ff, "r") as f:
+            # Read your local file here
+            # ...
+
+        There's no need to handle downloading of the file because it's on the local file system.
+        In this case, a dummy downloading will be done.
+
+        2. Directly open a FlyteFile with a remote path:
+
+        ff = FlyteFile(path=remote_path)
+        with open(ff, "r") as f:
+            # Read your remote file here
+            # ...
+
+        We now support directly opening a FlyteFile with a file from the remote data storage.
+        In this case, a delayed downloading of the remote file will be done.
+        For details, please refer to this issue: https://github.com/flyteorg/flyte/issues/6090.
+        """
+        ctx = FlyteContextManager.current_context()
+
+        if ctx.file_access.is_remote(self.path) and self._remote_source is None:
+            # Setup remote file source and local file destination
+            self._remote_source = self.path
+            local_path = ctx.file_access.get_random_local_path(self._remote_source)
+            self._downloader = lambda: FlyteFilePathTransformer.downloader(
+                remote_path=self._remote_source, local_path=local_path
+            )
+            self.path = local_path
```

Comment:
Would it be better to move these code to `__init__` ?

Reply:
Hi Kevin,

After some more tests, I observe that setting up these paths within `__init__` leads to an error. This use case is described as follows:
1. Initialize a `FlyteFile` with a remote path
2. Call `FlyteFilePathTransformer.async_to_literal` before `FlyteFile` is opened

The error shows that the local file destination to which we download the remote file hasn't been created because the we do **delayed downloading** until users call `open()` explicitly.

To be concrete, a non-existing `source_path` will be used [here](https://github.com/flyteorg/flytekit/blob/bc0e8c09758930da072fb228411b1b87f537e420/flytekit/types/file/file.py#L464). Hence, a `FileNotFoundError` occurs.

If there's any suggestion, please let me know. Thanks a lot!